### ISSUE TITLE
kubectl-opslevel overwrites user- provided values with null (always sends full service object)

### ISSUE THREAD

**Issue by MarcinGinszt** [OpsLevel/kubectl-opslevel#272]
Hello! 
We were using kubectl-opslevel since some time already, and lately we decided to update from `v2022.10.19` to latest version. 
We also had to update config version to "1.3.0" to make it work.

After running updated version, we noticed that behaviour changed- apparently, kubectl-opslevel now sends complete service object, even if fields are not set. This is destructive for fields set by UI and not by annotations.

For example, let's consider setting "system" through annotation:
```
system: '.metadata.annotations."app.uw.systems/system"'
```
This annotation before was setting `system` field for services that had this annotation specified, and ignoring it for others.
Now, services without this annotation specified are failing since system `null` doesn't exists. 

We can change the default `null` to `empty`, which will make the kubectl-opslevel work-
```
system: '.metadata.annotations."app.uw.systems/system" // empty'
```
However, fields that were set by users in UI, and are not set in Kubernetes configuration, are now being overridden to `""`. Previous version was ignoring them (since they are not set in configuration). There doesn't seem to be any workaround to fix that.

Is this expected behaviour? 
If yes- could we create parameter to ignore unset fields?

**Comment by rocktavious** [comment]
Hey @MarcinGinszt - We'll take a look at this as this is unexpected for me.

One thing that is weird is that you say you were using the `system` attribute but in the version `v2022.10.19` there was no ability to set system.  Here is the configurable fields as of that version - https://github.com/OpsLevel/kubectl-opslevel/blob/v2022.10.19/src/config/config.go#L22

We didn't add the `system` attribute until the 2024 versions.

We will go about testing the other fields like language and framework but if this is not happening for those fields and its only for the system field then that makes sense we likely missed this functionality when implementing that field.

**Comment by taimoorgit** [comment]
Hi @MarcinGinszt, thank you very much for filing this detailed bug report.

I was able to reproduce the issue today. In short, this ties to a bug related to [handling of failed JQ expressions.](https://github.com/OpsLevel/kubectl-opslevel/issues/36)

v2022 is not affected by this bug. If you would like to keep using k8s sync, the safest thing for you to do is to roll back to that version.

We are working on fixing this bug in v2024 and it is a high priority for us right now.

> However, fields that were set by users in UI, and are not set in Kubernetes configuration, are now being overridden to "". Previous version was ignoring them (since they are not set in configuration).

To clarify what the expected behaviour of the program is, 

1. If it detects an attribute that is not set/is set to a different value in OpsLevel, it should submit an update to populate/update that attribute.

2. If it detects something is not set in k8s, but is set in OpsLevel, it should be unaffected.

3. If an attribute is not defined in your config it should not be touched at all.

>Is this expected behaviour?

In short, no. The issue here is that when trying to perform (2) and (3) fields are being set to `"null"` as opposed to `null` and are sometimes being set to `""` instead of `null`.

Please let me know if there is anything else I can clarify.

**Comment by lboynton** [comment]
Also came across this issue. Our config is something like as follows:

```
version: "1.3.0"
service:
  import:
    - selector:
        apiVersion: apps/v1
        kind: Deployment
        excludes:
          - .metadata.namespace == "kube-system"
          - .metadata.annotations."opslevel.com/ignore"
      opslevel:
        name: .metadata.name
        aliases:
          - '"k8s:\(.metadata.name)-\(.metadata.namespace)"'
        tags:
          assign:
            - '{"imported": "kubectl-opslevel"}'
            - '{"kind": "Deployment"}'
            - '{"namespace": "\(.metadata.namespace)"}'
          create:
            - '{"environment": "${environment}"}'
```

This ends removing fields like `description`, `language`, `framework` and `product` even though we haven't configured it to set it in the config. For example this is a redacted log line from the pod:

```
[foo-service] Updated Service - Diff:
  &opslevel.Service{
        ApiDocumentPath: "",
-       Description:     "A description",
+       Description:     "",
        Framework:       "",
        HtmlURL:         "https://app.opslevel.com/services/foo-service",
        ServiceId:       {Id: "xxx", Aliases: {"foo-service", "k8s:foo-service-bar-namespace"}},
-       Language:        "Go",
+       Language:        "",
        Lifecycle:       {Alias: "generally_available", Description: "Service is supporting features available to be used by all custo"..., Id: "xxx", Index: 5, ...},
        ManagedAliases:  {"foo-service", "k8s:foo-service-bar-namespace"},
        ... // 6 identical fields
        Tags: &{Nodes: {{Id: "xxx", Key: "environment", Value: "development"} ...}, PageInfo: {Start: "MQ", End: "MTU"}, TotalCount: 15},
        Tier: {Alias: "tier_2", Description: "Customer-facing service or repository. Failure results in degrad"..., Id: "xxx", Index: 2, ...},
        Timestamps: opslevel.Timestamps{
                CreatedAt: {Time: s"2022-07-20 13:42:26.91772 +0000 UTC"},
-               UpdatedAt: iso8601.Time{Time: s"2023-07-19 13:27:30.683234 +0000 UTC"},
+               UpdatedAt: iso8601.Time{Time: s"2024-03-13 14:05:35.660184 +0000 UTC"},
        },
        Tools:        &{Nodes: {}},
        Dependencies: nil,
        ... // 2 identical fields
  }
```

**Comment by taimoorgit** [comment]
@lboynton Thank you for including your configuration file. This will help us with testing our next release that fixes then issue.

The next release is expected tomorrow, Thurs Mar 14th. It will prevent unwanted data overwrites that you both are experiencing. We are sorry about this problem, I know that it has been very disruptive. 

> Furthermore, this issue wasn't obvious from running a kubectl opslevel service preview command

This is unacceptable. 

We will address this in a release after the hotfix. We can make this `service preview` command more accurate by making it actually run a `service import` except it will just output the Service create/update requests instead of submitting them.

**Comment by lboynton** [comment]
Great, thanks @taimoorgit!

**Comment by lboynton** [comment]
No worries, thanks for the update!

**Comment by taimoorgit** [comment]
Hello @MarcinGinszt and @lboynton, we rolled out a release that addresses this bug: v2024.3.18. 

It will prevent fields belonging to a service (framework, language, description, name, tier, lifecycle, owner, system...) from getting overwritten as `"null"` or `""`.

We have written tests and manually tested to ensure that fields that are in a config but not in a deployment do not overwrite the data in web. The same applies to fields that are not present in a config. 

`config preview` has not changed, but we have a ticket to make it more accurate to what the reconciler actually does.

If you still face this issue, or any other issue where there's a chance of data loss please let us know ASAP. 

## Example

Kubernetes object (deployment.yaml):

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: demonstration
  labels:
    app: nginx
  annotations:
    opslevel.com/description: this was made for testing k8s sync
    opslevel.com/lifecycle: beta
    opslevel.com/owner: platform
    opslevel.com/tags.from_kubernetes: "True"
    opslevel.com/tags.hello: world
    opslevel.com/tags.image: "1.24"
    opslevel.com/tags.is_public: "True"
    opslevel.com/tags.vpc_enabled: "True"
    opslevel.com/tier: tier_2
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
        - name: nginx
          image: nginx:1.24
          ports:
            - containerPort: 8080
```

Output of running `service import` (with `config sample`):

```
5:31PM INF [apps/v1/deployments] Informer is ready and synced
5:31PM WRN unexpected happened: got service with alias 'default-demonstration' but the result has no ID
5:31PM WRN unexpected happened: got service with alias 'k8s:demonstration-default' but the result has no ID
5:31PM WRN [demonstration] Unable to find 'Lifecycle' with alias 'beta'
5:31PM INF [demonstration] Created new service
5:31PM INF [demonstration] Assigned alias 'default-demonstration'
5:31PM INF [demonstration] Assigned alias 'k8s:demonstration-default'
5:31PM INF [demonstration] Assigned tags: [{"key":"app","value":"nginx"},{"key":"from_kubernetes","value":"True"},{"key":"hello","value":"world"},{"key":"image","value":"1.24"},{"key":"imported","value":"kubectl-opslevel"},{"key":"is_public","value":"True"},{"key":"vpc_enabled","value":"True"}]
5:31PM INF [local-path-provisioner] No changes detected to fields - skipping update
5:31PM INF [local-path-provisioner] All tags already assigned to service.
5:31PM INF Import Complete
```

Now the service is created, in web when I change some fields:

<img width="826" alt="image" src="https://github.com/OpsLevel/kubectl-opslevel/assets/37668804/99c66aa0-7d5b-4301-8080-f51298a29378">

I run `service import` again with the sample config.

```
5:36PM INF [apps/v1/deployments] Informer is ready and synced
5:36PM WRN [demonstration] Unable to find 'Lifecycle' with alias 'beta'
5:36PM INF [demonstration] Detected Changes - Sending Update:
{"id":"Z2lkOi8vb3BzbGV2ZWwvU2VydmljZS8xMDg4Mjc","tierAlias":"tier_2"}
5:36PM INF [demonstration] Updated Service - Diff:
  &opslevel.Service{
  	... // 14 identical fields
  	Repositories: &{Edges: {}},
  	Tags:         &{Nodes: {{Id: "Z2lkOi8vb3BzbGV2ZWwvVGFnLzc5Njg3MDQy", Key: "app", Value: "nginx"}, {Id: "Z2lkOi8vb3BzbGV2ZWwvVGFnLzc5Njg3MDQz", Key: "from_kubernetes", Value: "True"}, {Id: "Z2lkOi8vb3BzbGV2ZWwvVGFnLzc5Njg3MDQ0", Key: "hello", Value: "world"}, {Id: "Z2lkOi8vb3BzbGV2ZWwvVGFnLzc5Njg3MDQ1", Key: "image", Value: "1.24"}, ...}, PageInfo: {Start: "MQ", End: "Nw"}, TotalCount: 7},
  	Tier: opslevel.Tier{
- 		Alias: "tier_4",
+ 		Alias: "tier_2",
  		Description: strings.Join({
- 			"Other",
+ 			"Customer-facing",
  			" services. Failure ",
- 			"doesn't result in immediate or significant impac",
  			... // 2 identical bytes
  		}, ""),
- 		Id:    "Z2lkOi8vb3BzbGV2ZWwvVGllci8yNDA",
+ 		Id:    "Z2lkOi8vb3BzbGV2ZWwvVGllci8yMzg",
- 		Index: 4,
+ 		Index: 2,
- 		Name:  "Tier 4",
+ 		Name:  "Tier 2",
  	},
  	Timestamps: opslevel.Timestamps{
  		CreatedAt: {Time: s"2024-03-18 21:31:06.104024 +0000 UTC"},
- 		UpdatedAt: iso8601.Time{Time: s"2024-03-18 21:35:12.770897 +0000 UTC"},
+ 		UpdatedAt: iso8601.Time{Time: s"2024-03-18 21:36:50.738445 +0000 UTC"},
  	},
  	Tools:        &{Nodes: {}},
  	Dependencies: nil,
  	... // 2 identical fields
  }

5:36PM INF [demonstration] All tags already assigned to service.
5:36PM INF [local-path-provisioner] No changes detected to fields - skipping update
5:36PM INF [local-path-provisioner] All tags already assigned to service.
5:36PM INF Import Complete
```

Notice here that only the tier was changed back to tier 2, because that was the only thing from the sample config that can be matched in the deploy data. (Lifecycle was not updated because the deployment.yaml doesn't have a valid lifecycle alias.)

In the previous v2024/v2023 versions, this would have resulted in system, product, owner, language, framework, lifecycle being set to `"null"` or `""`.
